hermes-client/src/main/java/pl/allegro/tech/hermes/client/HermesClient.java

```diff
@@ -67,16 +71,41 @@ private String createUri(URI uri) {
     }
 
     public CompletableFuture<HermesResponse> publish(HermesMessage message) {
+        if (shutdown) {
+            return completedWithShutdownException();
+        }
         HermesMessage.appendDefaults(message, defaultHeaders);
         return publish(message, (response) -> retryCondition.test(response) ? sendOnce(message) : completedFuture(response));
     }
 
     private CompletableFuture<HermesResponse> publish(HermesMessage message, Function<HermesResponse, CompletionStage<HermesResponse>> retryDecision) {
-        return range(0, retries).boxed().reduce(sendOnce(message), (future, attempt) -> future.thenCompose(retryDecision), (future, attempt) -> future);
+        currentlySending.incrementAndGet();
+        return range(0, retries).boxed().reduce(sendOnce(message), (future, attempt) -> future.thenCompose(retryDecision), (future, attempt) -> future)
+                .whenComplete((response, ex) -> currentlySending.decrementAndGet());
     }
 
     private CompletableFuture<HermesResponse> sendOnce(HermesMessage message) {
         return sender.send(URI.create(uri + message.getTopic()), message).exceptionally(HermesResponseBuilder::hermesFailureResponse);
     }
 
+    private CompletableFuture<HermesResponse> completedWithShutdownException() {
+        CompletableFuture<HermesResponse> alreadyShutdown = new CompletableFuture<>();
+        alreadyShutdown.completeExceptionally(new HermesClientShutdownException());
+        return alreadyShutdown;
+    }
+
+    public void shutdown() {
+        shutdown = true;
+    }
+
+    /**
+     * @return whether all inflight published messages have completed before timeout
+     */
+    public boolean awaitShutdown(long pollIntervalMillis, long timeout, TimeUnit timeoutTimeUnit) {
```

Comment:
Thanks for pointing this out, I'll remove thiss condition and simply await until the client is shutdown and the inflight counter is 0.